### PR TITLE
[ISSUE-52] - Remove Random Namespace

### PR DESCRIPTION
**Pull request by mtjandra**
* Addresses https://github.com/chef/inspec-aws/issues/52
* This should help make resources more identifiable
* Sample output (beginning and end) of running `bundle exec rake test:integration`
```
[inspec-aws (master)]> bundle exec rake test:integration
Please enter a namespace for your integration tests to run in: PullRequestDemo
----> Creating terraform environment
cd /home/mtjandra/code/inspec-aws/test/integration/build/ && terraform env new PullRequestDemo
Created and switched to environment "PullRequestDemo"!

You're now on a new, empty environment. Environments isolate their state,
so if you run "terraform plan" Terraform will not see any existing state
for this configuration.

...

Destroy complete! Resources: 3 destroyed.
----> Destroying terraform environment
cd /home/mtjandra/code/inspec-aws/test/integration/build/ && terraform env select default
Switched to environment "default"!
cd /home/mtjandra/code/inspec-aws/test/integration/build && terraform env delete PullRequestDemo
Deleted environment "PullRequestDemo"!
```

* Example of running individual tasks that now take parameters
```
[inspec-aws (master)]> bundle exec rake test:configure_test_environment[SingleTask]
----> Creating terraform environment
cd /home/mtjandra/code/inspec-aws/test/integration/build/ && terraform env new SingleTask
Created and switched to environment "SingleTask"!

You're now on a new, empty environment. Environments isolate their state,
so if you run "terraform plan" Terraform will not see any existing state
for this configuration.
[inspec-aws (master)]> bundle exec rake test:destroy_test_environment[SingleTask]
----> Destroying terraform environment
cd /home/mtjandra/code/inspec-aws/test/integration/build/ && terraform env select default
Switched to environment "default"!
cd /home/mtjandra/code/inspec-aws/test/integration/build && terraform env delete SingleTask
Deleted environment "SingleTask"!
```

### CODE REVIEW

**Review by Rugbyte — APPROVED** [review]

File: Rakefile [hunk 1]
```diff
@@ -63,18 +67,19 @@ namespace :test do
     sh("cd #{integration_dir}/build/ && terraform destroy -force")
   end
 
-  task :destroy_test_environment do
+  task :destroy_test_environment, :namespace do |t, args|
```
Comment (chris-redekop) [inline Rakefile]: This looks good.  I confirmed that I can call the **configure** and **destroy** tasks individually with the argument:

```
ubuntu@ubuntu-xenial:~/inspec-aws$ bundle exec rake test:destroy_test_environment[cr-foo]
----> Destroying terraform environment
cd /home/ubuntu/inspec-aws/test/integration/build/ && terraform env select default
Switched to environment "default"!
cd /home/ubuntu/inspec-aws/test/integration/build && terraform env delete cr-foo
Deleted environment "cr-foo"!
```

**Review by chris-redekop — APPROVED** [review]

**Comment by chris-rock** [comment]
I agree that we want to add a meaning to the namespace, but wouldn't it be better to prefix with a fixed name and use a random number then?

**Comment by mtjandra** [comment]
> I agree that we want to add a meaning to the namespace, but wouldn't it be better to prefix with a fixed name and use a random number then?

I believe we want to catch and fail on namespace conflicts, as this would suggest that the resources created from a previous Terraform run was not cleaned up correctly. Ideally we would want to catch these conflicting namespaces across all test runs, but the best we can do is detect conflicts based on some particular computer (terraform environments are not known to AWS).

**Comment by chris-redekop** [comment]
Yes, using a (partially) random ID to avoid env collisions will mask teardown issues from ANY previous runs.  I want to call this out as an anti-feature: the teardown failed for some reason, so we want to investigate and cleanup.  (Our experience has been that when teardown fails, resources are left with their AWS-cost meter running; and without an out-of-band process to clean these up, they and their costs just accumulate.)  For this reason, I propose that we don't mask collisions with a (partially) random ID.

OTOH, if you need to run multiple concurrent test setups, it's easy to manually tack on an incrementing integer to your env ID; so the lack of a (partially) random ID shouldn't be a blocker.

This would make doing the right thing--make sure your test resources are cleaned up--easier and doing the wrong thing--ignoring collisions with previous runs--harder.

What do you think, @chris-rock ?

**Review by aduric — APPROVED** [review]

**Review by vix633 — APPROVED** [review]

**Review by chris-rock — CHANGES_REQUESTED** [review]
While I understand your need and we should find a solution for this, I hesitate to approve a integration test that requires human interaction. There needs to be a path to use this in CI/CD environments. So how would I set the values in a CI environment that spins up terraform?

**Comment by chris-redekop** [comment]
Thanks, @chris-rock .  As discussed, an environment variable is a good approach here: if it exists, use its value as the Terraform env name; otherwise, prompt the user for a Terraform env name.

When I discussed with @mtjandra , he informed me that the tests already behave like this: they look for an environment variable named **INSPEC_TERRAFORM_ENV**.

I confirmed that this works as desired (no prompt!):

```bash
ubuntu@ubuntu-xenial:~/inspec-aws$ INSPEC_TERRAFORM_ENV=cr2 bundle exec rake test:integration
cd /home/ubuntu/inspec-aws/test/integration/build/ && terraform env new cr2
Created and switched to environment "cr2"!

You're now on a new, empty environment. Environments isolate their state,
so if you run "terraform plan" Terraform will not see any existing state
for this configuration.
----> Cleanup
cd /home/ubuntu/inspec-aws/test/integration/build/ && terraform destroy -force

Destroy complete! Resources: 0 destroyed.
----> Setup
cd /home/ubuntu/inspec-aws/test/integration/build/ && terraform plan
Refreshing Terraform state in-memory prior to plan...
The refreshed state will be used to calculate this plan, but will not be
persisted to local or remote state storage.
``` 
Is this what you meant, @chris-rock ?  If so, would you like to see any other changes here, or are we good to go?

File: Rakefile [hunk 2]
```diff
@@ -63,18 +67,19 @@ namespace :test do
...
     puts "----> Destroying terraform environment"
     sh("cd #{integration_dir}/build/ && terraform env select default")
-    sh("cd #{integration_dir}/build && terraform env delete #{terraform_env}")
+    sh("cd #{integration_dir}/build && terraform env delete #{args[:namespace]}")
   end
 
   task :integration do
-    Rake::Task["test:configure_test_environment"].execute
+    namespace = ENV['INSPEC_TERRAFORM_ENV'] || prompt("Please enter a namespace for your integration tests to run in: ")
```
Comment (chris-rock) [inline Rakefile]: Lets add documentation for that in our readme

Comment (chris-redekop) [inline Rakefile]: This made it into our README from the original "create separate env" issue.  Done!

Comment (chris-rock) [inline Rakefile]: Perfect, Thank you for double-checking

**Review by chris-rock — APPROVED** [review]
Nice improvement @mtjandra

**Pull request merged**